Optimize L1 MSR save/restore

ramesh-thomas opened the pull request:
Defer restoring L1 MSRs until there is a switch to user mode where the MSRs are used. This avoids unnecessary write to those MSRs if there was a switch to L2 guest while L1 is still in kernel mode.

This commit is applied over "mshv_vtl/tdx: Add assembly function for TDG.VP.ENTER" which has not been merged yet to avoid merge conflicts later when it is applied. Please merge https://github.com/microsoft/OHCL-Linux-Kernel/pull/54 first and ignore the TDG.VP.ENTER related commit in this PR.

ramesh-thomas committed:
mshv_vtl/tdx: Add assembly function for TDG.VP.ENTER

Replace inline assembly implementation with assembly function. This
makes the stack handling and the RBP save/restore cleaner. This is also
consistent with current upstream TDX implementations. The C function
that calls this function does not require the "noinline" directive with
this change.

Signed-off-by: Ramesh Thomas <ramesh.thomas@intel.com>

ramesh-thomas committed:
mshv_vtl/tdx: Optimize L1 MSR save/restore around VP_ENTER

Certain MSRs are not used in the L1 kernel mode and need to be restored
only when switching to user mode. It is possible that there would be a
switch to L2 guest before L1 switches to user mode. In that case the
MSRs are unnecessarily restored. Linux kernel provides a mechanism to
get notified when it is about to switch to user mode. The MSR
restoration is optimized by registering for that notification and
restoring them in the notifier callback.

Signed-off-by: Ramesh Thomas <ramesh.thomas@intel.com>

ramesh-thomas commented:
@ricardon

romank-msft commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,94 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#include <asm/asm-offsets.h>
```
I think there is an identical code in the other PR, can you tell why?

ramesh-thomas replied:
I had included the other patch to avoid merge conflicts when the other PR is merged because they modify same areas of code. I left a note in the description to merge PR #54 first and ignore the TDG.VP.ENTER commit included in this PR.  I can also remove the commit and rebase to top of tree after PR 54 is merged if that is easier.

romank-msft replied:
Thanks for explaining, either is good imo

chris-oo commented:
How was this tested?

ramesh-thomas commented:
> How was this tested?

Launched a L2 guest and traced the code path using debug prints.

chris-oo commented:
I think this needs a rebase after merging the other one.

ramesh-thomas commented:
> I think this needs a rebase after merging the other one.

Ok, I can rebase it.

chris-oo commented:
Actually, I merged your PR in the incorrect way that Allen is going to fixup. But i'll review this and it might need more fixup after main gets fixed, sorry about that.

chris-oo commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -801,42 +815,56 @@ void mshv_tdx_request_cache_flush(bool wbnoinvd)
+	per_cpu->urn_registered = false;
+	user_return_notifier_unregister(urn);
+
+	if (per_cpu->l1_msr_kernel_gs_base != per_cpu->l1_msr_kernel_gs_base_curr) {
```
how likely is it we actually hit this branch check where the L2 didn't have a different MSR value than what we expect? i'm wondering, should we just unconditionally restore the previous value if we enter usermode?

chris-oo replied:
IE I'm wondering, are we just always failing this check anyways, since we're going to restore the L2's MSRs before we dispatch?

ramesh-thomas replied:
I agree in our case we can restore them unconditionally. KVM does the checks but it has more complex conditions to handle e.g. shared OS etc. in which case the MSRs are likely to have not changed.

chris-oo commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -895,12 +894,17 @@ noinline void mshv_vtl_return_tdx(void)
+
+	if (!per_cpu->urn_registered) {
+		per_cpu->mshv_urn.on_user_return = mshv_vtl_on_user_return;
+		user_return_notifier_register(&per_cpu->mshv_urn);
```
i assume this callback is per-cpu?

ramesh-thomas replied:
The comment around the implementation suggests it is per-cpu. As MSR are per-cpu, the call needs to be on that cpu.
*
 * Request a notification when the current cpu returns to userspace.  Must be
 * called in atomic context.  The notifier will also be called in atomic
 * context.
 */

ramesh-thomas commented:
> Actually, I merged your PR in the incorrect way that Allen is going to fixup. But i'll review this and it might need more fixup after main gets fixed, sorry about that.

Ok, I will rebase to the main once it gets fixed.  While you are reviewing this, can you take a look at TSC_AUX MSR restoration? It is also getting restored when L1 switches to user mode. If it is used in L1 kernel then we would need to restore it immediately at VP.EXIT.

chris-oo commented:
I'm not sure about Linux kernel internals if TSC_AUX is used by the kernel itself. Let me ask around.

ramesh-thomas commented:
The original code was saving TSC_AUX differently than the other MSRs. It was getting saved right before VP.ENTER, which could imply that its value was expected to be changed after entering the kernel.  However, the comment at the place where it is getting saved suggests restoring it only when entering user mode.

ramesh-thomas commented:
@yamahata do you see any issue with deferring restoration of L1 TDX_AUX MSR to when user mode is entered instead of restoring immediately at VP.EXIT? The reasoning for deferring restoration of MSRs is the assumption that the MSRs are not getting used in the L1 kernel so it saves time if we restore only if there is a switch to L1 user mode.

yamahata commented:
> @yamahata do you see any issue with deferring restoration of L1 TDX_AUX MSR to when user mode is entered instead of restoring immediately at VP.EXIT? The reasoning for deferring restoration of MSRs is the assumption that the MSRs are not getting used in the L1 kernel so it saves time if we restore only if there is a switch to L1 user mode.

Yes we can. The VMX KVM and TDX KVM defers restoring TSC_AUX until returning back to user-mode (ring 3). We can follow them. If we know the user space doesn't use RDTSCP instruction, we can forcibly clear X86_FEATURE_RDTSCP for L1 kernel and skip restoring TSC_AUX.

ramesh-thomas commented:
> Yes we can. The VMX KVM and TDX KVM defers restoring TSC_AUX until returning back to user-mode (ring 3). We can follow them. If we know the user space doesn't use RDTSCP instruction, we can forcibly clear X86_FEATURE_RDTSCP for L1 kernel and skip restoring TSC_AUX.

A quick search in the Rust code found references to "rdtscp" but not sure if it actually gets used. I think we can leave it on and restore only when entering user mode.

benhillis commented:
> Actually, I merged your PR in the incorrect way that Allen is going to fixup. But i'll review this and it might need more fixup after main gets fixed, sorry about that.

@chris-oo & @allenpais - it should not be possible to merge PR's "the wrong way". Can we tweak the branch policies?

allenpais commented:
> > Actually, I merged your PR in the incorrect way that Allen is going to fixup. But i'll review this and it might need more fixup after main gets fixed, sorry about that.
> 
> @chris-oo & @allenpais - it should not be possible to merge PR's "the wrong way". Can we tweak the branch policies?

Yes, I am looking into it. Will do the needful.

allenpais commented:
> > > Actually, I merged your PR in the incorrect way that Allen is going to fixup. But i'll review this and it might need more fixup after main gets fixed, sorry about that.
> > 
> > 
> > @chris-oo & @allenpais - it should not be possible to merge PR's "the wrong way". Can we tweak the branch policies?
> 
> Yes, I am looking into it. Will do the needful.

Fixed.

ramesh-thomas commented:
@allenpais looks like the commit https://github.com/microsoft/OHCL-Linux-Kernel/pull/54/commits/d2d1830bb10d07dc709b418150c64620f027d9bc
in PR https://github.com/microsoft/OHCL-Linux-Kernel/pull/54 got removed.  That commit was included in this PR to avoid merge conflicts after #54 got merged.  That was also the cause of the confusion. 

Can you please remerge the commit in PR#54 and I will rebase this PR after removing the duplicate commit?

allenpais commented:
> @allenpais looks like the commit [d2d1830](https://github.com/microsoft/OHCL-Linux-Kernel/commit/d2d1830bb10d07dc709b418150c64620f027d9bc) in PR #54 got removed. That commit was included in this PR to avoid merge conflicts after #54 got merged. That was also the cause of the confusion.
> 
> Can you please remerge the commit in PR#54 and I will rebase this PR after removing the duplicate commit?

Yes, there were some changes merged by @chris-oo which needed a revert. I need to re-confirm which changes can go in. I will let you @chris-oo  comment on this change before I can merge. I hope that is okay.

ramesh-thomas commented:
@chris-oo @allenpais I have implemented the feedback and updated the commit.  Since this PR also contains the commit in PR#54 which has been reverted, we can now cleanly merge both the commits in this PR itself i.e. do not merge the one in #PR54 separately as I had suggested in my previous comment.

chris-oo commented:
Does this need to be rebased again to 6.12?

ramesh-thomas commented:
I can rebase them to 6.12.

ramesh-thomas commented:
@chris-oo done rebasing to 6.12

chris-oo reviewed (approved):
Changes LGTM.

ramesh-thomas commented:
Looks like the user-mode entry notifier is only supported for x86. Maybe the new code needs to be enclosed inside #ifdef CONFIG_X86_64. I will make the change and post a new patch.

ramesh-thomas commented:
@allenpais @chris-oo have resolved the ARM build issue and updated the commit.
